### PR TITLE
Clear dataloader references before attaching new dataloaders to Trainer

### LINKED ISSUE #8369: Trainer.validate(model, dataloader=dl) does not use passed dataloader
## 🐛 Bug

I wanted to validate my model on a subset of the original validation set. I found the Trainer.validate() function which takes one or multiple dataloaders as one of its arguments. However, the model will use the model's default dataloader and not the one I passed to .validate(). 

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1KMgKeugv9VqDgT4Mj8S-GObc8-3vS_H_?usp=sharing

### To Reproduce
Run the colab example. You will see that despite passing another dataloader to trainer.validate(), it still uses the initial dataloader.

### Expected behavior
If **no** new dataloader is passed to validate(), then it should use the existing dataloader of the model. 
If a new dataloader is passed to validate(), then the passed dataloader should be used. 

### Environment

 - PyTorch Lightning Version (e.g., 1.3.8):
 - PyTorch Version (e.g., 1.8)
 - Python version:
 - OS (e.g., Linux):
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - How you installed PyTorch (`conda`, `pip`, source):
 - If compiling from source, the output of `torch.__config__.show()`:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

- Fixes a bug introduced in #7207
- Part of #8435
- Related: #6030

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -1503,6 +1503,49 @@ def test_dataloaders_load_only_once_passed_loaders(tmpdir):
...
+    model = BoringModel()
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+
+    # 1st fit
+    trainer.fit(model, train_dataloaders=dataloader_0, val_dataloaders=dataloader_1)
+    assert trainer.train_dataloader.loaders is dataloader_0
+    assert trainer.val_dataloaders[0] is dataloader_1
+    # 2nd fit
+    trainer.fit(model, train_dataloaders=dataloader_2, val_dataloaders=dataloader_3)
+    assert trainer.train_dataloader.loaders is dataloader_2
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:1524]: So before this fix, the passing assertion would be?

```python
    assert trainer.train_dataloader.loaders is dataloader_1
```

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (awaelchli) [reply tests/trainer/test_dataloaders.py:1524]: before the fix, this assertion would pass: 
```python
trainer.fit(model, train_dataloaders=dataloader_0, val_dataloaders=dataloader_1)
# 2nd fit
trainer.fit(model, train_dataloaders=dataloader_2, val_dataloaders=dataloader_3)
assert trainer.train_dataloader.loaders is dataloader_0
```
the old dataloader remains attached to the Trainer.

**Review by ananthsub — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Great catch !
